File: frontend/src/app/services/customer.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Customer } from '../models/customer.model';
import { Page } from '../models/page.model';

@Injectable({
  providedIn: 'root'
})
export class CustomerService {
  private apiUrl = 'http://localhost:8080/api/customers';

  constructor(private http: HttpClient) { }

  getAllCustomers(page: number = 0, size: number = 25): Observable<Page<Customer>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<Page<Customer>>(this.apiUrl, { params });
  }

  getCustomerByCustomerId(customerId: number): Observable<Customer> {
    return this.http.get<Customer>(`${this.apiUrl}/${customerId}`);
  }

  getCustomerByEmail(emailId: string): Observable<Customer> {
    return this.http.get<Customer>(`${this.apiUrl}/email/${emailId}`);
  }

  createCustomer(customer: Customer): Observable<Customer> {
    return this.http.post<Customer>(this.apiUrl, customer);
  }

  updateCustomer(customerId: number, customer: Customer): Observable<Customer> {
    return this.http.put<Customer>(`${this.apiUrl}/${customerId}`, customer);
  }

  deleteCustomer(customerId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${customerId}`);
  }

  searchCustomers(name: string): Observable<Customer[]> {
    return this.http.get<Customer[]>(`${this.apiUrl}/search?name=${name}`);
  }

  getCustomersByCity(city: string): Observable<Customer[]> {
    return this.http.get<Customer[]>(`${this.apiUrl}/city/${city}`);
  }

  getCustomersByState(state: string): Observable<Customer[]> {
    return this.http.get<Customer[]>(`${this.apiUrl}/state/${state}`);
  }

  getCustomersByCommunicationPreference(preference: string): Observable<Customer[]> {
    return this.http.get<Customer[]>(`${this.apiUrl}/communication-preference/${preference}`);
  }
}

File: backend/src/main/java/com/teamcommercial/service/CustomerService.java
package com.teamcommercial.service;

import com.teamcommercial.entity.Customer;
import com.teamcommercial.repository.CustomerRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class CustomerService {

    private final CustomerRepository customerRepository;

    @Autowired
    public CustomerService(CustomerRepository customerRepository) {
        this.customerRepository = customerRepository;
    }

    public Page<Customer> getAllCustomers(Pageable pageable) {
        return customerRepository.findAll(pageable);
    }
    
    public List<Customer> getAllCustomers() {
        return customerRepository.findAll();
    }

    public Optional<Customer> getCustomerByCustomerId(Long customerId) {
        return customerRepository.findById(customerId);
    }

    public Optional<Customer> getCustomerByEmail(String emailId) {
        return customerRepository.findByEmailId(emailId);
    }

    public Customer createCustomer(Customer customer) {
        return customerRepository.save(customer);
    }

    public Customer updateCustomer(Long customerId, Customer customerDetails) {
        Customer customer = customerRepository.findById(customerId)
                .orElseThrow(() -> new RuntimeException("Customer not found with id: " + customerId));
        
        customer.setCustomerName(customerDetails.getCustomerName());
        customer.setCustomerId(customerDetails.getCustomerId());
        customer.setStreetNumber(customerDetails.getStreetNumber());
        customer.setStreetName(customerDetails.getStreetName());
        customer.setUnitNumber(customerDetails.getUnitNumber());
        customer.setCity(customerDetails.getCity());
        customer.setState(customerDetails.getState());
        customer.setPostCode(customerDetails.getPostCode());
        customer.setEmailId(customerDetails.getEmailId());
        customer.setMobileNumber(customerDetails.getMobileNumber());
        customer.setCommunicationPreference(customerDetails.getCommunicationPreference());
        customer.setSerial(customerDetails.getSerial());
        
        return customerRepository.save(customer);
    }

    public void deleteCustomer(Long customerId) {
        Customer customer = customerRepository.findById(customerId)
                .orElseThrow(() -> new RuntimeException("Customer not found with id: " + customerId));
        customerRepository.delete(customer);
    }

    public List<Customer> searchCustomersByName(String name) {
        return customerRepository.findByCustomerNameContainingIgnoreCase(name);
    }

    public List<Customer> getCustomersByCity(String city) {
        return customerRepository.findByCity(city);
    }

    public List<Customer> getCustomersByState(String state) {
        return customerRepository.findByState(state);
    }

    public List<Customer> getCustomersByCommunicationPreference(String preference) {
        return customerRepository.findByCommunicationPreference(preference);
    }
}

